test/smartwallet/CustomSmartWallet.test.ts

```diff
@@ -0,0 +1,339 @@
+import {
+  CustomSmartWalletInstance,
+  TestForwarderTargetInstance,
+  TestTokenInstance
+} from '../../types/truffle-contracts'
+
+// @ts-ignore
+import { EIP712TypedData, signTypedData_v4, TypedDataUtils } from 'eth-sig-util'
+import { BN, bufferToHex, privateToAddress, toBuffer } from 'ethereumjs-util'
+import { bytes32, containsEvent, getTestingEnvironment } from '../TestUtils'
+import TypedRequestData, { ForwardRequestType, getDomainSeparatorHash } from '../../src/common/EIP712/TypedRequestData'
+import { constants } from '../../src/common/Constants'
+import { RelayRequest } from '../../src/common/EIP712/RelayRequest'
+import RelayData from '../../src/common/EIP712/RelayData';
+import { expectRevert } from '@openzeppelin/test-helpers'
+import { ForwardRequest } from '../../src/common/EIP712/ForwardRequest'
+
+require('source-map-support').install({ errorFormatterForce: true })
+
+const TestForwarderTarget = artifacts.require('TestForwarderTarget')
+const TestToken = artifacts.require('TestToken')
+const TestSmartWallet = artifacts.require('TestSmartWallet')
+const SuccessCustomLogic = artifacts.require('SuccessCustomLogic')
+const FailureCustomLogic = artifacts.require('FailureCustomLogic')
+const ProxyCustomLogic = artifacts.require('ProxyCustomLogic')
+const CustomSmartWallet = artifacts.require('CustomSmartWallet')
+
+async function fillTokens (token: TestTokenInstance, recipient: string, amount: string): Promise<void> {
+  await token.mint(amount, recipient)
+}
+
+async function getTokenBalance (token: TestTokenInstance, account: string): Promise<BN> {
+  return await token.balanceOf(account)
+}
+
+function createRequest(request: Partial<ForwardRequest>, relayData: Partial<RelayData>): RelayRequest {
+   const baseRequest: RelayRequest = {
+    request: {
+      relayHub: constants.ZERO_ADDRESS,
+      from: constants.ZERO_ADDRESS,
+      to: constants.ZERO_ADDRESS,
+      value: '0',
+      gas: '1000000',
+      nonce: '0',
+      data: '0x',
+      tokenContract: constants.ZERO_ADDRESS,
+      tokenAmount: '1',
+      tokenGas: '50000'
+    },
+    relayData: {
+      gasPrice: '1',
+      domainSeparator: '0x',
+      relayWorker: constants.ZERO_ADDRESS,
+      callForwarder: constants.ZERO_ADDRESS,
+      callVerifier: constants.ZERO_ADDRESS
+    }
+  }
+  return { 
+    request: {
+      ...baseRequest.request,
+      ...request
+    }, relayData: {
+      ...baseRequest.relayData,
+      ...relayData
+    }
+  };
+}
+
+function signRequest(senderPrivateKey: Buffer, relayRequest: RelayRequest, chainId: number) {
+  const reqData: EIP712TypedData = new TypedRequestData(chainId, relayRequest.relayData.callForwarder, relayRequest)
+  const signature = signTypedData_v4(senderPrivateKey, { data: reqData })
+  const suffixData = bufferToHex(TypedDataUtils.encodeData(reqData.primaryType, reqData.message, reqData.types).slice((1 + ForwardRequestType.length) * 32))
+  return {signature, suffixData}
+}
```

Comment:
is there a source for this code?

Reply:
It was repeated in multiple tests. Check SmartWallet.test.ts

Reply:
ok, we can extract it later then. not the point of this PR :+1: